walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,78 @@
 
 package com.google.sps;
 
+import java.util.stream.Collectors;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
+import java.util.Set;
 
+/** 
+Algorithm:
+    1. Pick time ranges from events where attendees overlap with meeting request.
+    2. Now find the non overlapping time ranges which have duration more than request duration.
+*/
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    
+    private static final int END_OF_DAY = 24 * 60; 
+    
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {        
+        List<String> requestedAttendees = request.getAttendees().stream()
+                                                 .collect(Collectors.toList());
+        
+        ArrayList<TimeRange> overlappingRanges = calculateOverlappingRanges(events, requestedAttendees);
+        
+        long meetingDuration = request.getDuration();
+        
+        return getNonOverlappingTimeRanges(overlappingRanges, meetingDuration);
+    }
+
+    /** 1. Calculate overlapping ranges by iterating in all the events and finding common attendees.*/
+    private ArrayList<TimeRange> calculateOverlappingRanges(Collection<Event> events, List<String> requestedAttendees) {  
+        ArrayList<TimeRange> overlappingRanges = new ArrayList<>();
+
+        for (Event currentEvent: events) {
+            Set<String> currentEventAttendees = currentEvent.getAttendees();
+            List<String> overlapRange = currentEventAttendees.stream()
+                                                             .filter(requestedAttendees::contains)
+                                                             .collect(Collectors.toList());
+
+            if (overlapRange.size() > 0) {
+                overlappingRanges.add(currentEvent.getWhen());
+            }
+        }
```

Comment:
even this can be converted into a stream
currentEvent.stream().map().<current code>.collect.

Reply:
IMO converting `Event` to stream won't make any difference though.

Reply:
this is just to streamline the codebase into one semantic. I'd still suggest trying this as a java exercise later on :)